text/0000-io-os-reform.md

```diff
@@ -0,0 +1,1717 @@
+  platform-specific functionality and will be removed for now. They
+  may reappear later in `os::unix` and `os::windows` in a modified
+  form.
+* `dll_filename`: deprecated in favor of working directly with the constants.
+* `_NSGetArgc`, `_NSGetArgv`: these should never have been public.
+* `self_exe_path`: deprecated in favor of `current_exe` plus path operations.
+* `make_absolute`: deprecated in favor of explicitly joining with the working directory.
+* all `_as_bytes` variants: deprecated in favor of yielding `OsStrBuf` values
+
+### `std::fs`
+[std::fs]: #stdfs
+
+The `fs` module will provide most of the functionality it does today,
+but with a stronger cross-platform orientation.
+
+Note that all path-consuming functions will now take an
+`AsPath`-bounded parameter for ergonomic reasons (this will allow
+passing in Rust strings and literals directly, for example).
+
+#### Free functions
+[Free functions]: #free-functions
+
+**Files**:
+
+* `copy`. Take `AsPath` bound.
+* `rename`. Take `AsPath` bound.
+* `remove_file` (renamed from `unlink`). Take `AsPath` bound.
+
+* `file_attr` (renamed from `stat`). Take `AsPath` bound. Yield a new
+  struct, `FileAttr`, with no public fields, but `size`, `kind` and
+  `perm` accessors. The various `os::platform` modules will offer
+  extension methods on this structure.
+
+* `set_perm` (renamed from `chmod`). Take `AsPath` bound, and a
+  `FilePermissions` value. The `FilePermissions` type will be revamped
+  as a struct with private implementation; see below.
+
+**Directories**:
+
+* `make_dir` (renamed from `mkdir`). Take `AsPath` bound.
+* `make_dir_all` (renamed from `mkdir_recursive`). Take `AsPath` bound.
+* `read_dir` (renamed from `readdir`). Take `AsPath` bound. Yield a
+  newtypes iterator, which yields a new type `DirEntry` which has an
+  accessor for `Path`, but will eventually provide other information
+  as well (possibly via platform-specific extensions).
+* `remove_dir` (renamed from `rmdir`). Take `AsPath` bound.
+* `remove_dir_all` (renamed from `rmdir_recursive`). Take
+  `AsPath` bound.
+* `walk_dir`. Take `AsPath` bound. Yield an iterator over `IoResult<DirEntry>`.
+
+**Links**:
+
+* `hard_link` (renamed from `link`). Take `AsPath` bound.
+* `sym_link` (renamed from `symlink`). Take `AsPath` bound.
+* `read_link` (renamed form `readlink`). Take `AsPath` bound.
+
+#### Files
+[Files]: #files
+
+The `File` type will largely stay as it is today, except that it will
+use the `AsPath` bound everywhere.
+
+The `stat` method will be renamed to `attr`, yield a `FileAttr`, and
+take `&self`.
+
+The `fsync` method will be renamed to `flush_os`, and `datasync` will
+be moved to `os::unix` (since it has no meaning on Windows)
+
+The `path` method wil remain `#[unstable]`, as we do not yet want to
+commit to its API.
+
+The `open_mode` function will take an `OpenOptions` struct, which will
+encompass today's `FileMode` and `FileAccess` and support a
+builder-style API.
```

Comment:
It would be great if we could get a way to atomically open a file only if there already isn't a file at that location (a facility needed for temporary files etc. and which, at least according to the documentation, windows and unix support equally well). I put together a sketch of this relative to the current library at https://github.com/ltratt/rust/commit/b20467d7c94ffb554244d05b1f22e63adfd90018